pkg/virt-controller/watch/vm.go

```diff
@@ -1748,7 +1754,61 @@ func (c *VMController) isVirtualMachineStatusStopped(vm *virtv1.VirtualMachine,
 
 // isVirtualMachineStatusStopped determines whether the VM status field should be set to "Provisioning".
 func (c *VMController) isVirtualMachineStatusProvisioning(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) bool {
-	return !c.areDataVolumesReady(vm) || !c.arePVCVolumesReady(vm)
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
+		if volume.DataVolume == nil {
+			continue
+		}
+
+		dv, err := c.getDataVolumeFromCache(vm.Namespace, volume.DataVolume.Name)
+		if err != nil {
+			log.Log.Object(vm).Errorf("Error fetching DataVolume while determining virtual machine status: %v", err)
+			continue
+		}
+		if dv == nil {
+			continue
+		}
+
+		// Skip DataVolumes with unbound PVCs since these cannot possibly be provisioning
+		dvConditions := controller.NewDataVolumeConditionManager()
+		if !dvConditions.HasConditionWithStatus(dv, cdiv1.DataVolumeBound, k8score.ConditionTrue) {
+			continue
+		}
+
+		if dv.Status.Phase != cdiv1.Succeeded {
+			return true
+		}
+	}
+
+	return false
+}
+
+// isVirtualMachineStatusWaitingForVolumeBinding
+func (c *VMController) isVirtualMachineStatusWaitingForVolumeBinding(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) bool {
+	if !isSetToStart(vm, vmi) {
+		return false
+	}
+
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
+		claimName := typesutil.PVCNameFromVirtVolume(&volume)
+		if claimName == "" {
+			continue
+		}
+
+		pvc, err := c.getPersistentVolumeClaimFromCache(vm.Namespace, claimName)
+		if err != nil {
+			log.Log.Object(vm).Errorf("Error fetching PersistentVolumeClaim while determining virtual machine status: %v", err)
+			continue
+		}
+		if pvc == nil {
+			continue
+		}
+
+		if pvc.Status.Phase != k8score.ClaimBound {
+			return true
+		}
```

Comment:
does waiting for first consumer need to be accounted for here?

Reply:
Actually no, since the `WaitingForVolumeBinding` status is considered only when the VM is set to start:

```go
func (c *VMController) isVirtualMachineStatusWaitingForVolumeBinding(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) bool {
    if !isSetToStart(vm, vmi) {
        return false
    }
        ...
```

Reply:
Actually, there is some time period in which the VM is set to start, and the virt-launcher pod hasn't been scheduled yet. In this case, from the perspective of the storage subsystem, there's still no consumer for the PVC. But I think this is negligible enough and we can report that (typically very short) time period as `WaitingForVolumeBinding` too. If we ever add a `Scheduling` status, this can replace it then.

Reply:
ah, i see. So the running status will match the printable status first in this case. I think this is all still accurate even with WFFC.